### ISSUE TITLE
Unreproducible jpm sign api issue

### ISSUE THREAD

**Issue by Noitidart** [mozilla/olympia#1369]
I was using the JPM Sign API and did this:

		var promise_sign = xhr('https://addons.mozilla.org/api/v3/addons/THIS_CAN_BE_ANYTHING/versions/' + aAddonVersionInXpi + '/', {
			method: 'PUT',
			data: formData,
			responseType: 'json',
			headers: {
				Authorization: 'JWT ' + generateToken(aAmoApiKey, aAmoApiSecret)
			}
		});

And the response of the status check:

				var promise_signStatus = xhr(aCheckSignedURL, {
					responseType: 'json',
					headers: {
						Authorization: 'JWT ' + generateToken(aAmoApiKey, aAmoApiSecret)
					}
				});

was this:

![](file:///C:/Users/Mercurius/Documents/jpm%20sign%20status%20when%20not%20yet%20processed.png)

I kept running that check for about 30 minutes, checking every second, and it would never change.

Checking AMO we see status is marked "incomplete".

I cannot reproduce this. Does anyone know what happend? And how I can detect status was incomplete and the signing will never happen?

Here is the addon on amo - https://addons.mozilla.org/en-US/developers/addon/two-bootstrap-nsitimer/edit

**Comment by kumar303** [comment]
My best guess is that we hit an intermittent deadlock or other DB error and it rolled back your add-on submission. I don't know of a way to check for this currently because the 'incomplete' state is the default state. You could give yourself a timeout to give up after. As for the stability issues (sorry), pretty much the whole team is working on smoothing these out as we speak. Tracker: https://github.com/mozilla/olympia/issues/1244

**Comment by kumar303** [comment]
Since we're tracking API performance in a separate issue and we don't have a great idea for making the API respond better to catastrophic failure, I'm closing.

**Comment by Noitidart** [comment]
Ah can those stability be issues be responsible for me getting `status == 401`, `statusText == "Unauthorized"` and `response.detail == 'Incorrect authentication credentials.'`

I tried changing the timing in my token generator and all kinds of things but randomly I still get these. If it's on the stability issues that might make sense because I can't explain it haha

**Comment by andymckay-limited-access** [comment]
I don't think so, most of the problems are occurring on writes. Do you have
any more details on those requests?

**Comment by Noitidart** [comment]
Thanks @andymckay-limited-access for the quick reply. Here is my full code, it works, but randomly I get those invalid errors, so I have a timeout to try it again and again till it works.

https://github.com/Noitidart/Chrome-Store-Foxified/blob/fbfadd28835d5cc30e9baa44774793047fde6ba5/bootstrap.js#L57-L241

This is my jsonwebtoekn, I just ran `browserify` on the node jsonwebtoken module -

https://github.com/Noitidart/Chrome-Store-Foxified/blob/fbfadd28835d5cc30e9baa44774793047fde6ba5/modules/jsonwebtoken.js

**Comment by andymckay-limited-access** [comment]
Sorry more interested in any HTTP logs for those transactions (please blank
out your API keys).

**Comment by Noitidart** [comment]
Ah. Here is http to the sign call.

```
https://addons.mozilla.org/api/v3/addons/Bootstrap-nsITimer-EIGHT%40jetpack/versions/0.3/

PUT /api/v3/addons/Bootstrap-nsITimer-EIGHT%40jetpack/versions/0.3/ HTTP/1.1
Host: addons.mozilla.org
User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64; rv:44.0) Gecko/20100101 Firefox/44.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Authorization: JWT eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJ1c2VyOjEyMDg0MTYyOjQ1NCIsImp0aSI6IjAuNDYyNzc3MDA5MzU2NzAzMiIsImlhdCI6MTQ1MjgxMzkyNSwiZXhwIjoxNDUyODEzOTg1fQ.mVME4u5IXZaJxinXRrYl8rRQCODh_QcTGWpPQsYY42M
Content-Length: 3431
Content-Type: multipart/form-data; boundary=---------------------------191261131116310
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
-----------------------------191261131116310
Content-Disposition: form-data; name="Content-Type"

multipart/form-data
-----------------------------191261131116310
Content-Disposition: form-data; name="upload"; filename="nsi2.xpi"
Content-Type: application/x-xpinstall

PK

HTTP/1.1 401 UNAUTHORIZED
Allow: GET, PUT, HEAD, OPTIONS
Content-Type: application/json
Date: Thu, 14 Jan 2016 23:25:15 GMT
Server: nginx
Set-Cookie: multidb_pin_writes=y; expires=Thu, 14-Jan-2016 23:25:30 GMT; Max-Age=15; Path=/
Strict-Transport-Security: max-age=31536000
Vary: Accept, X-Mobile, User-Agent
WWW-Authenticate: JWT realm="api"
X-Frame-Options: DENY
Transfer-Encoding: chunked
Connection: keep-alive
```

**Comment by kumar303** [comment]
If you are only getting the 401 Unauthorized response intermittently then, yes, it is likely a timing problem. I would suggest trying to sync your clock with [NTP](http://www.ntp.org/) or some equivalent. Let us know if that doesn't fix it.

**Comment by Noitidart** [comment]
@kumar303 wow that's sensitive. Thanks! Right after sync I get intermittent, but less then 10 min later Im getting it again, re-re-syncing fixed it again. Is this a bug? Or is it sensitive by design?

**Comment by kumar303** [comment]
it sounds like you have a drift in your clock. Our site allows for a 5 second drift but I guess we could increase it. Did you try using something like NTP to keep it constantly in sync?

**Comment by Noitidart** [comment]
Thanks @kumar303  for your replies. The issue is definitely my side (client side), my clock. :)

##### tldr:
Is that 5 second allowance here to stay? Or are their plans to make it not so time dependent?

If the time sensitivity stays, do we have any sure fire techniques to ensure time is on target?

##### Reasons why I ask:
I'm worried about this time sensitivity because I'm landing this in my addon, which takes extensions on chrome webstore, turns it into an xpi (appends `-foxified` to their id), signs it with this api, and installs it into firefox. So users of this addon will be calling the API.

If their clocks are out of sync, it's an issue that needs troubleshooting, which is bad on my part as the dev of that addon.

If the time sensitivity stays, I have brainstormed two techqniues to handle it, does the team have any recommendations to people who run into this? These are my ideas:

1. Get a server, sync that server to NTP every so often, setup webpage that my addon can hit, on load it gives a time (or token). Then addon uses that as credentials.
2. Prompt the users on failed credentials, and ask for permission to sync their system clocks, and do it for them. (i dont think the average addon installer of chrome addons knows how to sync clocks haha)

Thanks for the replies, I know we're talking about a topic other then the original post, really appreciate it!

**Comment by kumar303** [comment]
> Is that 5 second allowance here to stay? Or are their plans to make it not so time dependent?

by nature, [JWTs](https://jwt.io/introduction/) are time dependent. We could increase the 5 second allowance if needed but that doesn't sound like a perfect solution here.

> Does the AMO API offer a page I can hit to get a proper time that I can use to calculate a token...?

It doesn't but we could add something like this. I'd suggest filing a separate issue for that.

> I'm landing this in my addon, which takes extensions on chrome webstore, turns it into an xpi (appends -foxified to their id), signs it with this api, and installs it into firefox. So users of this addon will be calling the API.

This suggests to me that you are exposing your secret API key in your add-on code because that's the only way users could make signed requests to the API on your behalf (or are they expected to configure their own API credentials?). Either way, the API currently isn't designed to be used from an unprotected environment such as a browser. We plan to support this environment but will need to implement something more like Oauth for it to be secure.

You could consider creating an intermediate server that the add-on can use to safely proxy API requests. On the server you can run NTP to guarantee its clock is in sync when creating JWTs. You have to be very careful with this approach, however, because if you allow any arbitrary source code to be signed with your API credentials through the proxy, then malware could be traced back to you (which you probably don't want).

**Comment by kumar303** [comment]
One way to make the proxy secure is to let the caller only provide an ID of the Chrome store extension. The proxy server would then fetch code securely from the Chrome store before converting and signing.

**Comment by Noitidart** [comment]
@kumar303 thanks for those super cool ideas!! I like that server doing all the heavy lifting idea!
